### PR TITLE
Added code to enable EventList support for Matter

### PR DESCRIPTION
**Pull request by emargolis**
- Added new method to extract events from database: `selectEndpointClusterEvents()`
 - Added new helper functions to construct list (array) of supported event IDs:
     `chip_endpoint_generated_event_count()`
     `chip_endpoint_generated_event_list()`
 - Added `eventList` and `eventCount` entries to the EmberAfCluster struct in the generated `endpoint_config.h` file

The EventList data is treated similar to CommandList data.

The assumption is that events are always `server --> client`.

### CODE REVIEW

File: src-electron/util/zcl-util.js [hunk 1]
```diff
@@ -75,6 +75,23 @@ function commandComparator(a, b) {
...
+ * @param {*} a
+ * @param {*} b
+ * @returns -1, 0 or 1
+ */
+function eventComparator(a, b) {
+  if (a.manufacturerCode < b.manufacturerCode) return -1
+  if (a.manufacturerCode > b.manufacturerCode) return 1
+
+  if (a.hexCode < b.hexCode) return -1
+  if (a.hexCode > b.hexCode) return 1
```
Comment (bzbarsky-apple) [inline src-electron/util/zcl-util.js:90]: So here, sort by code, not hexcode, once we switch to the existing shared event mapping?

File: test/gen-template/matter/endpoint-config.zapt [hunk 2]
```diff
@@ -37,6 +37,13 @@
 
 #define ZAP_GENERATED_COMMANDS_INDEX(index) ((chip::CommandId *) (&generatedCommands[index]))
 
+// clang-format off
+#define GENERATED_EVENT_COUNT {{ chip_endpoint_generated_event_count }}
+#define GENERATED_EVENTS {{ chip_endpoint_generated_event_list }}
+// clang-format on
+
+#define ZAP_GENERATED_EVENTS_INDEX(index) ((chip::EventId *) (&generatedEvents[index]))
```
Comment (bzbarsky-apple) [inline test/gen-template/matter/endpoint-config.zapt:45]: Why do we need a cast here?  I don't think we should have a cast here....  We should just have our types be actually right.  Why don't we?

Comment (bzbarsky-apple) [reply test/gen-template/matter/endpoint-config.zapt:45]: I know we have it for commands, but we should not have it there either, either here or on the Matter side.

Comment (emargolis) [reply test/gen-template/matter/endpoint-config.zapt:45]: Yes, it should work without casting. Just copied if from the line above for `generatedCommands`

File: src-electron/db/query-endpoint.js [hunk 3]
```diff
@@ -265,6 +265,51 @@ ORDER BY C.CODE
...
+  E.EVENT_ID = ETE.EVENT_REF
+WHERE
+  E.CLUSTER_REF = ?
+  AND ETE.ENDPOINT_TYPE_REF = ?
+ORDER BY E.MANUFACTURER_CODE, E.CODE
+  `,
+    [clusterId, endpointTypeId]
+  )
+
+  return rows.map((row) => {
```
[the remark this answers is not in the document]
Comment (emargolis) [reply src-electron/db/query-endpoint.js]: updated

File: src-electron/util/zcl-util.js (hunk 1 above)
Comment (emargolis) [reply src-electron/util/zcl-util.js]: done

**Comment by emargolis** [comment]
@bzbarsky-apple Thanks for a quick review. Your comments are addressed.

File: src-electron/db/query-endpoint.js [hunk 4]
```diff
@@ -265,6 +265,40 @@ ORDER BY C.CODE
...
+async function selectEndpointClusterEvents(db, clusterId, endpointTypeId) {
+  return dbApi.dbAll(
+    db,
+    `
+SELECT
+  E.EVENT_ID,
+  E.NAME,
+  E.CODE,
+  E.MANUFACTURER_CODE,
+  E.IS_OPTIONAL
```
Comment (bzbarsky-apple) [inline src-electron/db/query-endpoint.js:285]: You might want to select more columns here; see the set of things that `dbMapping.map.event` maps.

**Review by bzbarsky-apple — APPROVED** [review]

File: src-electron/db/query-endpoint.js (hunk 4 above)
Comment (emargolis) [reply src-electron/db/query-endpoint.js]: updated

**Review by robszewczyk — APPROVED** [review]

**Pull request merged**